File: examples/hangman/c/hangman.c
#include "hangman.h"

WordHint words[NUM_WORDS] = {
    {"python", "A programming language known for its readability."},
    {"gravity", "A natural force that pulls objects toward each other."},
    {"castle", "A large fortified building from medieval times."},
    {"keyboard", "A device with keys used to type."},
    {"forest", "A large area covered mostly with trees."},
    {"planet", "A celestial body orbiting a star."},
    {"battery", "A portable power source."},
    {"mountain", "A large natural elevation of the Earth's surface."},
    {"library", "A place where books are kept."},
    {"rhythm", "A repeated pattern of movement or sound."}
};

// Hangman drawings
const int NUM_PICS = 7;
const char *hangman_pics[] = {
    "\n"
    "     +---+\n"
    "         |\n"
    "         |\n"
    "         |\n"
    "       ===",

    "\n"
    "     +---+\n"
    "     O   |\n"
    "         |\n"
    "         |\n"
    "       ===",

    "\n"
    "     +---+\n"
    "     O   |\n"
    "     |   |\n"
    "         |\n"
    "       ===",

    "\n"
    "     +---+\n"
    "     O   |\n"
    "    /|   |\n"
    "         |\n"
    "       ===",

    "\n"
    "     +---+\n"
    "     O   |\n"
    "    /|\\  |\n"
    "         |\n"
    "       ===",

    "\n"
    "     +---+\n"
    "     O   |\n"
    "    /|\\  |\n"
    "    /    |\n"
    "       ===",

    "\n"
    "     +---+\n"
    "     O   |\n"
    "    /|\\  |\n"
    "    / \\  |\n"
    "       ==="
};


int choose_word(WordHint *selected) {
	int index = rand() % NUM_WORDS;
	*selected = words[index];
	return 0;
}

void display_game_state(GameState *state) {
	printf("\n%s\n", hangman_pics[ state->wrong_count ]);
	printf("\n");

	// secret status
	printf("Word: ");
	char *secret = state->secret;
	for (int i=0; i< state->len_secret; ++i) {
		if ( char_already_guessed(state, secret[i])) {
			printf(" %c", secret[i]);
		} else {
			printf(" _");
		}
	}

	// guessed letters
	//    - letters[] buffer holds ready-to-print texts.
	//      Which includes ' ' for spacing.  So, size is 2x of
	//      actual buffer size.
	char letters[ NUM_ALPHABET * 2 ];
	list_guessed_letters(state, letters);
	printf("\nGuessed letters: %s\n\n", letters);
}

// Parameter:
//   - state:   GameState, as usual
//   - letters: (buf: allocated from callee),
//              directly printable for convenience.
// Returns:  0 for success,
//           non-zero for reserve
//
int list_guessed_letters(GameState *state, char *letters) {
	char sorted[NUM_ALPHABET];
	memset(sorted, '\0', NUM_ALPHABET);

	// Copy guessed letters to sorted array
	memcpy(sorted, state->guessed, state->guess_count);

	// Bubble Sort.
	//   - Consider: For efficient Set() operation,
	//          BST might fit better.
	//   - For now, lazy method.
	for (int i = 0; i < state->guess_count - 1; ++i) {
		for (int j = 0; j < state->guess_count - i - 1; ++j) {
			if (sorted[j] > sorted[j + 1]) {
				char temp = sorted[j];
				sorted[j] = sorted[j + 1];
				sorted[j + 1] = temp;
			}
		}
	}

	// Build output string with spaces between letters
	int pos = 0;
	for (int i = 0; i < state->guess_count; ++i) {
		if (i > 0) {
			letters[pos++] = ' ';
		}
		letters[pos++] = sorted[i];
	}
	letters[pos] = '\0';  // Null-terminate the string

	return 0;
}

bool char_in_secret(GameState *state, char chk) {
	char *secret = state->secret;
	for (int i=0; i< state->len_secret; ++i)
		if (secret[i] == chk)
			return true;
	return false;
}

bool char_already_guessed(GameState *state, char chk) {
	char *guessed = state->guessed;
	for (int i=0; i< state->guess_count; ++i)
		if (guessed[i] == chk)
			return true;
	return false;
}

bool all_guessed(GameState *state) {
	char *secret = state->secret;
	for (int i=0; i< state->len_secret; ++i) 
		if ( ! char_already_guessed(state, secret[i]))
			return false;
	return true;
}

bool game_lost(GameState *state) {
	return state->wrong_count >= NUM_PICS;
}

File: examples/hangman/python/hangman.py
"""
Hangman CLI game.
Good Sample Output:  from Gemini's prototype

```
=== HANGMAN ===
Guess the word one letter at a time.
Type 'hint' for a definition, 'quit' to exit.


     +---+
         |
         |
         |
       ===

Word: _ _ _ _ _ _
Guessed letters:

Your guess:
```

"""


import os, sys
import random
from typing import Tuple

# Word list with hints
words = [
    ("python", "A programming language known for its readability."),
    ("gravity", "A natural force that pulls objects toward each other."),
    ("castle", "A large fortified building from medieval times."),
    ("keyboard", "A device with keys used to type."),
    ("forest", "A large area covered mostly with trees."),
    ("planet", "A celestial body orbiting a star."),
    ("battery", "A portable power source."),
    ("mountain", "A large natural elevation of the Earth's surface."),
    ("library", "A place where books are kept."),
    ("rhythm", "A repeated pattern of movement or sound.")
]

# Hangman drawings
hangman_pics = [
    """
     +---+
         |
         |
         |
       ===""",

    """
     +---+
     O   |
         |
         |
       ===""",

    """
     +---+
     O   |
     |   |
         |
       ===""",

    """
     +---+
     O   |
    /|   |
         |
       ===""",

    """
     +---+
     O   |
    /|\\  |
         |
       ===""",

    """
     +---+
     O   |
    /|\\  |
    /    |
       ===""",

    """
     +---+
     O   |
    /|\\  |
    / \\  |
       ==="""
]

def choose_word() -> Tuple[str, str]:
    return random.choice(words)

class Hangman:
    def __init__(self):
        self.secret, self.hint = choose_word()
        self.guessed = []
        self.wrong: int = 0

    def display_game_state(self):
        print( hangman_pics[self.wrong])
        print( f"\nWord: {self.secret_spaced()}")
        print( f"Guessed letters: {' '.join(sorted(self.guessed))}")

    def secret_spaced(self) -> str:
        lst = []
        for sc in self.secret:    # sc := secret_char
            if sc in self.guessed:
                lst.append(sc)
            else:
                lst.append("_")
        return ' '.join(lst)

    def win(self) -> bool:
        for sc in self.secret:
            if sc not in self.guessed:
                return False
        return True
    
    def lost(self) -> bool:
        return self.wrong >= len(hangman_pics)
    

def main():
    print(" == HANGMAN Game ==")

    # Init
    game = Hangman()
    while not (game.win() or game.lost()):
        game.display_game_state()
        try:
            user_input = input("\nYour guess: ")
        except EOFError:
            # Handle Ctrl-D
            break

        user_input = user_input.strip()
        if len(user_input) != 1:
            print("I said: SINGLE CHAR!!")
            continue

        guess = user_input[0].lower()
        if guess in game.guessed:
            print(f"Char({guess}) is already guessed. Try again.")
            continue

        game.guessed.append(guess)
        if guess not in game.secret:
            game.wrong += 1

    if game.win():
        print("Congratulations!  You won.")
    elif game.lost():
        print(f"Sorry.  The word was: {game.secret}")
    else:
        #unlikely
        print("Should not reach here.")
        

if __name__ == '__main__':
    main()
